src/QMCWaveFunctions/Fermion/MultiSlaterDeterminantFast.cpp

```diff
@@ -265,31 +266,36 @@ WaveFunctionComponent::PsiValueType MultiSlaterDeterminantFast::evalGrad_impl_no
                                                                                             bool newpos,
                                                                                             GradType& g_at)
 {
-  const bool upspin = getDetID(iat) == 0;
-  const int spin0   = (upspin) ? 0 : 1;
-  const int spin1   = (upspin) ? 1 : 0;
+  const int det_id = getDetID(iat);
 
   if (newpos)
-    Dets[spin0]->evaluateDetsAndGradsForPtclMove(P, iat);
+    Dets[det_id]->evaluateDetsAndGradsForPtclMove(P, iat);
   else
-    Dets[spin0]->evaluateGrads(P, iat);
+    Dets[det_id]->evaluateGrads(P, iat);
 
-  const GradMatrix_t& grads            = (newpos) ? Dets[spin0]->new_grads : Dets[spin0]->grads;
-  const ValueType* restrict detValues0 = (newpos) ? Dets[spin0]->new_detValues.data() : Dets[spin0]->detValues.data();
-  const ValueType* restrict detValues1 = Dets[spin1]->detValues.data();
-  const size_t* restrict det0          = (*C2node)[spin0].data();
-  const size_t* restrict det1          = (*C2node)[spin1].data();
+  const GradMatrix_t& grads            = (newpos) ? Dets[det_id]->new_grads : Dets[det_id]->grads;
+  const ValueType* restrict detValues0 = (newpos) ? Dets[det_id]->new_detValues.data() : Dets[det_id]->detValues.data();
+  const size_t* restrict det0          = (*C2node)[det_id].data();
   const ValueType* restrict cptr       = C->data();
   const size_t nc                      = C->size();
-  const size_t noffset                 = Dets[spin0]->FirstIndex;
+  const size_t noffset                 = Dets[det_id]->FirstIndex;
   PsiValueType psi(0);
   for (size_t i = 0; i < nc; ++i)
   {
     const size_t d0 = det0[i];
     //const size_t d1=det1[i];
     //psi +=  cptr[i]*detValues0[d0]        * detValues1[d1];
     //g_at += cptr[i]*grads(d0,iat-noffset) * detValues1[d1];
-    const ValueType t = cptr[i] * detValues1[det1[i]];
+    ValueType t = cptr[i];
+    for (size_t id = 0; id < Dets.size(); id++)
+    {
+      if (id != det_id)
+      {
+        const ValueType* restrict detValues1 = Dets[id]->detValues.data();
+        const size_t* restrict det1          = (*C2node)[id].data();
+        t *= detValues1[det1[i]];
```

Comment:
Could you avoid detValues1 and det1 variables and write the code compact.

Reply:
Done